File: Problems/LeetCode/C457/B.py
import collections
from sortedcontainers import SortedList

class DSU:
    def __init__(self, n):
        self.fa = list(range(n))
    def find(self, o):
        if self.fa[o] == o:
            return o
        self.fa[o] = self.find(self.fa[o])
        return self.fa[o]
    def merge(self, u, v):
        u = self.find(u)
        v = self.find(v)
        if u != v:
            self.fa[u] = v
class Solution:
    def processQueries(self, n: int, e: List[List[int]], Q: List[List[int]]) -> List[int]:
        g = DSU(n + 1)
        for u, v in e:
            g.merge(u, v)

        st = collections.defaultdict(SortedList)
        for i in range(1, n + 1):
            f = g.find(i)
            st[f].add(i)

        ans = []
        for op, x in Q:
            f = g.find(x)
            if op == 1:
                if len(st[f]) > 0:
                    if x in st[f]:
                        ans.append(x)
                    else:
                        ans.append(st[f][0])
                else:
                    ans.append(-1)
            elif len(st[f]) > 0 and x in st[f]:
                    st[f].remove(x)
        return ans

File: Problems/LeetCode/C457/B.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef double db;
typedef pair<int,int> PII;
typedef unsigned long long ull;
const int inf = 1000000000;

vector<int> fa;
int find(int o) { return o == fa[o] ? fa[o] : fa[o] = find(fa[o]); }
void merge(int u, int v) {
    u = find(u), v = find(v);
    if(u == v) return;
    fa[u] = v;
}
//懒删除堆写法
class Solution {
public:
    vector<int> processQueries(int n, vector<vector<int>>& e, vector<vector<int>>& Q) {
        fa.assign(n + 1, 0);
        iota(fa.begin(), fa.end(), 0);
        for(int i = 0; i < e.size(); i++) {
            int u = e[i][0], v = e[i][1];
            merge(u, v);
        }
        vector<priority_queue<int, vector<int>, greater<int>>> q(n + 1);
        for(int i = 1; i <= n; i++) {
            int f = find(i);
            q[f].push(i);
        }
        vector<int> ans;
        vector<bool> vis(n + 1);
        for(int i = 0; i < Q.size(); i++) {
            int op = Q[i][0], x = Q[i][1];
            int f = find(x);
            if(op == 1) {
                if(!vis[x]) {
                    ans.push_back(x);
                    continue;
                }
                auto &t = q[f];
                while(!t.empty() && vis[t.top()]) {
                    t.pop();
                }
                if(!t.empty()) {
                    ans.push_back(t.top());
                } else {
                    ans.push_back(-1);
                }
            } else {
                vis[x] = true;
            }
        }
        return ans;
    }
};




// class Solution {
// public:
//     vector<int> processQueries(int n, vector<vector<int>>& e, vector<vector<int>>& Q) {
//         fa.assign(n + 1, 0);
//         iota(fa.begin(), fa.end(), 0);
//         for(int i = 0; i < e.size(); i++) {
//             int u = e[i][0], v = e[i][1];
//             merge(u, v);
//         }
//         unordered_map<int,set<int>> st;
//         for(int i = 1; i <= n; i++) {
//             int f = find(i);
//             st[f].insert(i);
//         }
//         int q = Q.size();
//         vector<int> ans;
//         for(int i = 0; i < q; i++) {
//             int op = Q[i][0], x = Q[i][1];
//             int f = find(x);
//             if(op == 1) {
//                 if(!st[f].empty()) {
//                     if(st[f].count(x)) ans.push_back(x);
//                     else ans.push_back(*st[f].begin());
//                 } else {
//                     ans.push_back(-1);
//                 }
//             } else {
//                 st[f].erase(x);
//             }
//         }
//         return ans;
//     }
// };